File: Dante2/Module2.05/Zadanie5.03/main.c
#include <stdio.h>
#include <stdlib.h>

int create_array_2d_2(int ***ptr, int width, int height);
void destroy_array_2d(int ***ptr, int height);
void display_array_2d(int **ptr, int width, int height);
int sum_array_2d(int **ptr, int width, int height);
int sum_row(int *ptr, int width);

int main()
{
    int error;

    int width, height;
    printf("Enter: ");

    error = scanf("%i %i",&width, &height);
    if(error != 2)
    {
        printf("Incorrect input");
        return 1;
    }
    if(height <= 0 || width <= 0)
    {
        printf("Incorrect input data");
        return 2;
    }

    int **ptr;
    error = create_array_2d_2(&ptr, width, height);
    if(error == 1)
    {
        printf("Incorrect input");
        return 1;
    }
    if(error == 2)
    {
        printf("Failed to allocate memory");
        return 8;
    }

    printf("Enter:\n");
    for(int i = 0; i < height; i++)
    {
        for(int j = 0; j < width; j++)
        {
            error = scanf("%i",(*(ptr + i) + j));
            if(!error)
            {
                destroy_array_2d(&ptr, height);
                printf("Incorrect input");
                return 1;
            }
        }
    }
    display_array_2d(ptr, width, height);

    for(int i = 0; i < height; i++)
    {
        printf("%i\n", sum_row(*(ptr + i), width));
    }
    printf("%i", sum_array_2d(ptr, width, height));

    destroy_array_2d(&ptr, height);
    return 0;
}

int sum_array_2d(int **ptr, int width, int height)
{
    if(ptr == NULL || width <= 0 || height <= 0)
    {
        return -1;
    }

    int suma = 0;

    for(int i = 0; i < height; i++)
    {
        for(int j = 0; j < width; j++)
        {
            suma = suma + *(*(ptr + i) + j);
        }
    }

    return suma;
}

int sum_row(int *ptr, int width)
{
    if(ptr == NULL || width <= 0)
    {
        return -1;
    }

    int suma = 0;

    for(int i = 0; i < width; i++)
    {
        suma = suma + *(ptr + i);
    }

    return suma;
}

int create_array_2d_2(int ***ptr, int width, int height)
{
    if(width <= 0 || height <= 0 || ptr == NULL)
    {
        return 1;
    }

    int **array = malloc(sizeof(int *) * height);
    if(array == NULL)
    {
        return 2;
    }

    for(int i = 0; i < height; i++)
    {
        *(array + i) = malloc(sizeof(int) * width);
        if(*(array + i) == NULL)
        {
            destroy_array_2d(&array, height);
            return 2;
        }
    }

    *ptr = array;

    return 0;
}

void display_array_2d(int **ptr, int width, int height)
{
    if (ptr == NULL || width <= 0 || height <= 0)
    {
        return;
    }

    for (int i = 0; i < height; i++)
    {
        for (int j = 0; j < width; j++)
        {
            printf("%i ", *(*(ptr + i) + j));
        }
        printf("\n");
    }
}

void destroy_array_2d(int ***ptr, int height)
{
    if (ptr == NULL || height <= 0)
    {
        return;
    }

    int **temp = *ptr;
    for (int i = 0; i < height; i++)
    {
        free(*(temp + i));
    }
    free(temp);

    *ptr = NULL;
}

File: Dante2/Module2.05/Zadanie5.03/main.rs
use regex::Regex;

fn create_array_2d(width: i32, height: i32) -> Result<Vec<Vec<i32>>, String> {
    if width <= 0 || height <= 0 {
        return Err("Width and height must be positive integers.".to_string());
    }

    let array: Vec<Vec<i32>> = vec![vec![0; width as usize]; height as usize];

    Ok(array)
}

fn display_array_2d(array: & Vec<Vec<i32>>, width: i32, height: i32) {
    if width < 0 || height < 0 {
        return;
    }

    println!();
    for i in 0..height {
        for j in 0..width {
            print!("{} ", array[i as usize][j as usize]);
        }
        println!();
    }
    println!();
}

fn sum_array_2d(array: & Vec<Vec<i32>>, width: i32, height: i32) -> i32 {
    if width <= 0 || height <= 0 {
        return -1;
    }

    let mut sum: i32 = 0;

    for i in 0..height {
        for j in 0..width {
            sum = sum + array[i as usize][j as usize];
        }
    }

    sum
}
fn sum_row(array: & Vec<i32>, width: i32) -> i32 {
    if width <= 0 {
        return -1;
    }

    let mut sum: i32 = 0;

    for i in 0..width {
        sum = sum + array[i as usize];
    }

    sum
}

fn main() {
    let mut input: String = String::new();
    println!("Enter: ");
    std::io::stdin().read_line(&mut input).expect("Incorrect input");
    let re = Regex::new(r"(\d+)\s+(\d+)").expect("Incorrect REGEX");
    if let Some(caps) = re.captures(&input) {
        let width: i32 = caps[1].parse().expect("Incorrect width parse");
        let height: i32 = caps[2].parse().expect("Incorrect height parse");

        if width <= 0 || height <= 0 {
            panic!("Incorrect input data");
        }

        match create_array_2d(width, height) {
            Ok(mut array) => {
                println!("Enter: ");

                for i in 0..height {
                    for j in 0..width {
                        input.clear();
                        std::io::stdin().read_line(&mut input).expect("Incorrect input in array");
                        array[i as usize][j as usize] = input.trim().parse().expect("Incorrect parsing");
                    }
                }

                display_array_2d(&array, width, height);

                for i in 0..height {
                    println!("Column: {}, sum is: {}", i, sum_row(& array[i as usize], width));
                }
                println!();

                println!("Total sum is: {}", sum_array_2d(&array, width, height));
            }
            Err(e) => eprintln!("Error: {}", e),
        }
    }
    else {
        panic!("Incorrect numbers");
    }
}
